### DISCUSSION TITLE
collection('something').delete() does not recursively delete nested collections

### DISCUSSION THREAD

**Discussion by lockieluke** [invertase/react-native-firebase#6281] [Ideas] [answered]
there's a function for recursive delete in the official firestore library i do think we should add this to the library

**Comment by mikehardy** [comment, marked as answer]
Unfortunately that's not the way this has to work because deleting recursively requires running firestore operations in large batches (because there is a 500 document limit to deletes, and the recursive operation may total more than that...), and it's unreliable in a mobile context. It has to be done in a callable function, which is not something we can directly package for people.

But firebase has provided ample documentation and sample code for this case, and I use it myself in my apps. It works great

This is what you want

- https://firebase.google.com/docs/firestore/solutions/delete-collections
- https://github.com/firebase/snippets-node/tree/master/firestore/solution-deletes

Integrate that to your functions implementation and call it from your app as needed

**Comment by lockieluke** [comment]
```ts
deleteCollectionRecursively: (collection: FirebaseFirestoreTypes.CollectionReference<FirebaseFirestoreTypes.DocumentData>) => {
    return new Promise<void>(async (resolve, reject) => {
        const [err, data] = await to(collection.get());
        if (err) reject(err);
        if (data?.docs.length)
            await Promise.all(data.docs.map(doc => collection.doc(doc.id).delete()));

        resolve();
    })
}
```

dont worry mate i just found the solution it works quite well

**Reply by mikehardy** [reply]
That's not actually recursive (there may be sub-collections), and if you have more than 500 documents it will throw an error.

It would more appropriately be named `deleteUpTo500DocumentsInThisCollectionWhileOrphaningSubCollectionsOrError()`

https://firebase.google.com/docs/firestore/solutions/delete-collections#solution_delete_data_with_a_callable_cloud_function

```
Deleting entire collections from a resource-limited mobile app can be difficult to implement for the following reasons:

- There is no operation that atomically deletes a collection.
- Deleting a document does not delete the documents in its subcollections.
- If your documents have dynamic subcollections, it can be hard to know what data to delete for a given path.
- Deleting a collection of more than 500 documents requires multiple batched write operations or hundreds of single deletes.
- In many apps, it isn't appropriate to give end-users permission to delete entire collections.
```

**Reply by lockieluke** [reply]
right thanks for the reminder

**Reply by lockieluke** [reply]
i can probably modify the function to just delete 500 at a time and use recursion to delete the rest of it, and eventually call resolve when its completed

**Reply by mikehardy** [reply]
If you do this on the client, it will be unreliable due to failing network connections etc, and could consume quite a lot of battery as it goes back and forth to do the deletes. The only recommended solution is the one I've linked in the documents above - you make a callable function using firebase functions, and just call the function from the client. I do that in my apps (as does everyone I think?) and it works great

**Reply by lockieluke** [reply]
im on the free plan lol should i just host a server on heroku to do it

**Reply by lockieluke** [reply]
since im using the collections for user comments i could just limit the number of comments to 500 for now until my app becomes a bit more popular

**Reply by mikehardy** [reply]
Interesting, I didn't realize that functions wasn't included in the free plan at all. Worth noting that if you go to the blaze plan you get all the free plan stuff, plus for functions the first 2 million calls a month are free. I have a small app and I find myself paying about $2 a month total, and that's only because I do some extensive firestore work for full export / processing. Not saying that would be a good idea for your or not, just to say that it may actually be $0 to upgrade

The biggest fear for me as an app developer is unexpected virality (and a big non-monetized bill...). 

You might like item 7 from the "how to cap your firebase costs" official youtube series https://www.youtube.com/playlist?list=PLl-K7zZEsYLmK1tiMBeKA0iDMPDCJKM-5

**Reply by lockieluke** [reply]
I can't activate the blaze plan cuz I don't have a credit card

**Reply by mikehardy** [reply]
Definitely a show stopper there :laughing: - alright then, doing it on device might be your best option, or using firebase-admin sdk any other spot as you mentioned. Good luck with your project

**Reply by lockieluke** [reply]
Thanks mate!

**Comment by aitmamatovmusa** [comment]
In the node js environment, you can delete nested documents and collections via `recursiveDelete`
```ts
import { firestore } from "firebase-admin";

await firestore().recursiveDelete(collectionRef || documentRef);
```
